When `edges` command tests fail print all pods for additional debugging information.

kleimkuhler opened the pull request:
After #6574 recently merging the `TestDirectEdges` test has been flaky. Occasionally the test fails for the following error:
```
2021-08-05T22:08:06.2860090Z --- FAIL: TestDirectEdges (73.39s)
2021-08-05T22:08:06.2860719Z     edges_test.go:161: Expected output:
2021-08-05T22:08:06.2861172Z         \[
2021-08-05T22:08:06.2861480Z           \{
2021-08-05T22:08:06.2861884Z             "src": "prometheus",
2021-08-05T22:08:06.2862642Z             "src_namespace": "external\-prometheus",
2021-08-05T22:08:06.2863356Z             "dst": "slow-cooker",
2021-08-05T22:08:06.2864211Z             "dst_namespace": "linkerd-direct-edges-test",
2021-08-05T22:08:06.2865267Z             "client_id": "prometheus.external\-prometheus",
2021-08-05T22:08:06.2866374Z             "server_id": "default.linkerd-direct-edges-test",
2021-08-05T22:08:06.2867120Z             "no_tls_reason": ""
2021-08-05T22:08:06.2867492Z           \},
2021-08-05T22:08:06.2867804Z           \{
2021-08-05T22:08:06.2868211Z             "src": "prometheus",
2021-08-05T22:08:06.2868933Z             "src_namespace": "external\-prometheus",
2021-08-05T22:08:06.2869481Z             "dst": "terminus",
2021-08-05T22:08:06.2870311Z             "dst_namespace": "linkerd-direct-edges-test",
2021-08-05T22:08:06.2871318Z             "client_id": "prometheus.external\-prometheus",
2021-08-05T22:08:06.2872435Z             "server_id": "default.linkerd-direct-edges-test",
2021-08-05T22:08:06.2873180Z             "no_tls_reason": ""
2021-08-05T22:08:06.2873563Z           \},
2021-08-05T22:08:06.2873888Z           \{
2021-08-05T22:08:06.2874421Z             "src": "slow-cooker",
2021-08-05T22:08:06.2875282Z             "src_namespace": "linkerd-direct-edges-test",
2021-08-05T22:08:06.2875936Z             "dst": "terminus",
2021-08-05T22:08:06.2876751Z             "dst_namespace": "linkerd-direct-edges-test",
2021-08-05T22:08:06.2877906Z             "client_id": "default.linkerd-direct-edges-test",
2021-08-05T22:08:06.2879151Z             "server_id": "default.linkerd-direct-edges-test",
2021-08-05T22:08:06.2879887Z             "no_tls_reason": ""
2021-08-05T22:08:06.2880258Z           \}
2021-08-05T22:08:06.2880579Z         \]
2021-08-05T22:08:06.2880885Z         
2021-08-05T22:08:06.2881227Z         actual:
2021-08-05T22:08:06.2881560Z         [
2021-08-05T22:08:06.2881876Z           {
2021-08-05T22:08:06.2882429Z             "src": "slow-cooker",
2021-08-05T22:08:06.2883268Z             "src_namespace": "linkerd-direct-edges-test",
2021-08-05T22:08:06.2883925Z             "dst": "terminus",
2021-08-05T22:08:06.2884759Z             "dst_namespace": "linkerd-direct-edges-test",
2021-08-05T22:08:06.2885901Z             "client_id": "default.linkerd-direct-edges-test",
2021-08-05T22:08:06.2887153Z             "server_id": "default.linkerd-direct-edges-test",
2021-08-05T22:08:06.2887896Z             "no_tls_reason": ""
2021-08-05T22:08:06.2888250Z           }
2021-08-05T22:08:06.2888572Z         ]
2021-08-05T22:08:06.2888870Z
```

What is happening is Prometheus has not scraped workloads yet, so there is no edge to or from Prometheus and slow-cooker.

While it's remains unclear why exactly this happens, this change addresses two potential problems:

1. Print the cluster's pods so that we can determine if the Prometheus pod is injected. If it's not injected, we'll be able to debug further with that information.
2. Introduce retries to the `TestEdges` test. As we already have retries in the `TestDirectEdges`, this makes a similar change to `TestEdges`.

Signed-off-by: Kevin Leimkuhler <kevin@kleimkuhler.com>

alpeb reviewed (approved):
LGTM :+1:

adleong reviewed (commented):
I don't understand how the "Expected Prometheus edges" error is functionally different from the "Expected output:" error.  In either case, I would expect the test to be retried up to the timeout.  How does this help the flakeyness?

kleimkuhler commented:
Hm that's a good point I guess we're already retrying the `Expected output ...` error. I was trying to avoid introducing a sleep, but maybe that's our only option here.

A 10 second sleep at the start of the function that we pass into the retry may be best.

alpeb commented:
@kleimkuhler The last time I tried to address this flakiness was in #6575. The problem was that the Prometheus pod wasn't being injected. Maybe that's still the case? Given this has been such a persistent issue, it might be worth adding more debugging info when this test fails, and leave that debugging there for good. Like showing the output of `kubectl get po -A` just to see if the pods are injected.

kleimkuhler commented:
If that's still the case, I've added `kubectl get pods -A` output to the `Expected output ...` error. I removed the sleep I introduced so that we have a better chance of reproducing this and can see what the cluster's pod state is.

adleong reviewed (approved):
Listing the pods in the error message seems fine but we should just be careful when interpreting this information.  There will be a period of time when the Prometheus pod exists but has not yet scraped and created any edges.  Prometheus pods which exist but are not running (for example if they have been evicted) will also show here, I believe; which could be confusing.

kleimkuhler commented:
Yep all good points. I think the main thing we'll be able to takeaway as Alejandro mentioned is if the Prometheus pod is getting injected or not.

By the way, I added retries to the `TestEdges` test because it actually failed on the previous commit for this lack of edges reason. I'm not sure if that was obvious if you had not had a chance to look at the previous CI failure.
